### ISSUE TITLE
Warning: "Selector xx is never used"

### ISSUE THREAD

**Issue by GarryFlemings** [timo-reymann/intellij-mjml-support#43]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Used versions

- Windows 10 Home 20H2 19042.1083
- WebStorm 2021.1.3 #WS-211.7628.25 built on June 30, 2021
- MJML Support 2021.07.09-322

## Condition

I have an email with two external CSS files. I use `mj-include` within `mj-head` to get each CSS file into the project. The first CSS file defines these and only these CSS `selector-list`s, all via `mj-style`:

```
body
h1, h2
h1
h2
li, ol, ul
li.liPosOne
li.liPosTwo
p
table, tbody, th, tr, td
.mj-carousel-main
.mj-carousel-main > tbody > tr > td:first-of-type 
.mj-carousel-main > tbody > tr > td:nth-of-type(3)
.mj-carousel-previous
.mj-carousel-next
.mj-carousel-thumbnail
.mj-carousel-image-1
```

The second CSS file defines these and only these CSS `selector-list`s, all via `mj-style`:

```
a
a.a
h1, h2
.responsive
.accordionImg
.tdOuterAcc
.tdAcc
.styleCenter
.colDemTabN
.colDemTabW
.colDemTr
.colDemTd1
.colDemTd2
.colDemTd3
.tableAcc
table#versions
table#versions a
table#versions > tr > td
table#versions > tr > td.otherVersions
table#versions td.web
table#versions td.plain
table.banner
table.banner tr > td.bannerLogo
table.banner tr > td.bannerText
table.banner td.bannerText > h1.WIN
table.banner p.WINcontact > a
```

The same email also has  several external MJML-only files.  I use `mj-include` within `mj-body` to get each of these into the project. Among them, these files use all of the above `selector-list`s (or parts of them) _except_

In the first CSS file, not used:

```
ol
li.liPosTwo
th
```

[The seven `.mj-carousel*` classes in the first file are not used in my MJML. MJML defines them in the HTML it generates for `mj-carousel`; they are used there.]

In the second CSS file, not used:

```
.responsive
```

### Actual Behavior

With small horizontal lines in the editor and on the "Problems" tab (not in the "Event Log"), I see these reports.

![Screenshot 2021-07-10 164259a](https://user-images.githubusercontent.com/14095071/125179757-0d19ee00-e1b7-11eb-97c1-ae4e529466c3.png)

As to the second report there, "Unknown attribute line-height for mjml tag mj-accordion-text": not in scope to this write-up.

### Steps to Reproduce

Working on my email.

### Important Factoids

About reports in the editor and Problem Tab:

- There are three potential reports for `selectors-list`s in the first CSS file.
- No report is about a `selectors-list` in the first CSS file. (Not preferred.)

- These is one potential report for the second CSS file (i.e., `.responsive`). It is among the reports. (Preferred.)
- All other reports are for `selectors-list`s used in the code. (Not preferred.)
- No report is about a `selectors-list` using a selector (e.g., `table#versions > tr > td`).
- Some reports are about a part of a `selectors-list` using a selector (e.g., `web` for `table#versions td.web`).
- The last report (attributed to line 350) refers to the line `table.banner p.WINcontact > a` and reports on only part of the related identifier.

**Comment by timo-reymann** [comment]
Currently the plugin only provided references to css selectors defined implicitly by mjml components.

As for includes/inline blocks this seems like a really useful feature to me. But i need to do some research how IntelliJ is handling it currently for e.g. HTML.

This might take some time, and i will keep this issue updated :)

**Comment by timo-reymann** [comment]
As the plugin will need more contextual information, i will need to add at least one index for "memorizing" all used css classes inside mjml code with their occurence.

With that information it will be possible to tell the IDE that the according classes have been used.

Currently i am still in the prototyping of an index for this and the implementation. As i have very limited experience with indices this takes some time.

**Comment by timo-reymann** [comment]
Current state:
- Usage is working (IDE is no longer complaining + highlights it)
- reference to css files works
- reference to mj-style block works (without refactoring support -> hard to figure out why, css is not open source)
- 3 new indices (mj-included files, classes defined in mj-style blocks, mj-class usage)

**Comment by timo-reymann** [comment]
UPDATE: The mj-style is currently implemented using injection for css, the only way known to me for using css inside xml without implementing a whole language.

Thats also the reason why the ref doesnt work for mj-style blocks. For css file includes it works fine

**Comment by GarryFlemings** [comment]
I just noticed 2021.08.03-356 and retested this. I have worked a bit on this email between tests, so this is not fully "apples-to-apples", but I think this is no concern in this case.

With small horizontal lines in the editor and In the "Problems" tab (the same places mentioned above), __none of the reports pictured above are present__. The "Unknown attribute line-height for mjml tag mj-accordion-text"--which I specified as not in scope to this write-up--is absent, too. A large majority of the conditions generating reports above indeed remain present in the existing test.

🎉🙏
